lib/tables.c

```diff
@@ -4675,3 +4675,146 @@ table_collection_simplify(table_collection_t *self,
+
+static int table_cleaner_run(table_cleaner_t *self)
+{
+    // Remove any sites with duplicate positions, retaining only the *first*
+    // one. Assumes the tables have been sorted, throwing an error if not.
+    // TODO: this could be done in place rather than by constructing a new
+    // table.
+
+    int ret = 0;
+
+    if (self->sites->num_rows <= 1) {
+        // nothing to do
+        goto out;
+    }
+
+    table_size_t j, k, site_j;
+    table_size_t as_length, as_offset;
+    table_size_t md_length, md_offset;
+    double last_position, position;
+
+    self->site_id_map[0] = 0;
+    site_j = 0;
+    last_position = self->sites->position[0];
+    as_offset = self->sites->ancestral_state_offset[1];
+    md_offset = self->sites->metadata_offset[1];
+    for (j = 1; j < self->sites->num_rows; j++) {
+        position = self->sites->position[j];
+        if (position < last_position) {
+            ret = MSP_ERR_UNSORTED_SITES;
+            goto out;
+        }
+        if (position != last_position) {
+            site_j++;
+            as_length = self->sites->ancestral_state_offset[j + 1] - self->sites->ancestral_state_offset[j];
+            md_length = self->sites->metadata_offset[j + 1] - self->sites->metadata_offset[j];
+            if (site_j != j) {
+                assert(site_j < j);
+                self->sites->position[site_j] = self->sites->position[j];
+                memcpy(self->sites->ancestral_state + self->sites->ancestral_state_offset[site_j],
+                        self->sites->ancestral_state + self->sites->ancestral_state_offset[j],
+                        as_length);
+                self->sites->ancestral_state_offset[site_j] = as_offset;
+                memcpy(self->sites->metadata + self->sites->metadata_offset[site_j],
+                        self->sites->metadata + self->sites->metadata_offset[j],
+                        md_length);
+                self->sites->metadata_offset[site_j] = md_offset;
+            }
+            as_offset += as_length;
+            md_offset += md_length;
+            last_position = position;
+        }
+        self->site_id_map[j] = site_j;
+    }
+
+    if (site_j == j) {
+        // we haven't changed anything, no need to remap sites
+        goto out;
+    }
+
+    // reset site table size
+    self->sites->num_rows = site_j + 1;
+    self->sites->ancestral_state_length = self->sites->ancestral_state_offset[site_j + 1];
+    self->sites->metadata_length = self->sites->metadata_offset[site_j + 1];
+
+    // Remap sites in the mutation table
+    for (j = 0; j < self->mutations->num_rows; j++) {
+        k = self->mutations->site[j];
+        assert(k <- self->sites->num_rows);
+        self->mutations->site[j] = self->site_id_map[k];
+    }
+
+out:
+    return ret;
+}
+
+static int 
+table_cleaner_alloc(table_cleaner_t *self, 
+        site_table_t *sites,
+        mutation_table_t *mutations) {
+    int ret = 0;
+
+    memset(self, 0, sizeof(table_cleaner_t));
```

Comment:
This is perfect. Just to clarify the reasoning for setting the struct to zero first. This is to make sure that if an error happens during initialisation, and pointers within the struct will be set to NULL, and so ``msp_safe_free`` will work properly. If we don't do this, then any errors during ``_alloc`` will lead to a crash during ``_free``.

Reply:
That's exactly why I had that in there -- before I had it, I got a crash during `free`. =)